Add webhook function to ingress custom events (#606)
vmware-samples/vcenter-event-broker-appliance: examples/knative/powershell/kn-ps-webhook/README.md

# kn-ps-webhook

> Example Knative PowerShell function that accepts a custom incoming webhook and automatically constructs a CloudEvent and forwards that to the VMware Event Router.

Reviewer: ```suggestion
Example Knative PowerShell function that accepts a custom incoming webhook,  constructs a CloudEvent and forwards it to the VMware Event Router (`broker`).
```

![](screenshots/screenshot-01.png)

## Prerequisites

> * VEBA deployed with a [wildcard DNS entry](https://en.wikipedia.org/wiki/Wildcard_DNS_record), which is required for the ability to externally invoke a Knative Service URL that is serving the custom webhook function. For more information, please [refer to this blog post](https://williamlam.com/2021/09/custom-webhook-function-to-publish-events-into-vmware-event-broker-appliance-veba.html) for more information.

Reviewer: ```suggestion
* VEBA deployed with a [wildcard DNS entry](https://en.wikipedia.org/wiki/Wildcard_DNS_record), which is required for the ability to externally invoke a Knative Service URL that is serving the custom webhook function. Please [refer to this blog post](https://williamlam.com/2021/09/custom-webhook-function-to-publish-events-into-vmware-event-broker-appliance-veba.html) for more information.
```

# Step 1 - Build

Create the container image locally to test your function logic.

```console
export TAG=<version>
docker build -t <docker-username>/kn-ps-webhook:${TAG} .
```

# Step 2 - Test

Verify the container image works by executing it locally.

Change into the `test` directory
```console
cd test
```

Update the following variable names within the `docker-test-env-variable` file and either the `send-cloudevent-test.ps1` (PowerShell Script) or `send-cloudevent-test.sh` (Bash Script)

* WEBHOOK_USERNAME - Username to login to webhook function endpoint
* WEBHOOK_PASSWORD - Password to login to webhook function endpoint

Start the container image by running the following command:

```console
docker run -e FUNCTION_DEBUG=true -e PORT=8080 --env-file docker-test-env-variable -it --rm -p 8080:8080 <docker-username>/kn-ps-webhook:${TAG}
```

> In a separate terminal, run either `send-cloudevent-test.ps1` (PowerShell Script) or `send-cloudevent-test.sh` (Bash Script) to simulate a CloudEvent payload being sent to the local container image

Reviewer: nit: You don't send CE payload here

```console
Testing Function ...
See docker container console for output

# Output from docker container console
Id     Name            PSJobTypeName   State         HasMoreData     Location             Command
--     ----            -------------   -----         -----------     --------             -------
2      Job2            ThreadJob       NotStarted    False           PowerShell           …
09/19/2021 13:34:04 - PowerShell HTTP server start listening on 'http://*:8080/'
09/19/2021 13:34:04 - Processing Init

09/19/2021 13:34:04 - Init Processing Completed

09/19/2021 13:34:09 - Processing Context ...

09/19/2021 13:34:09 - [Function Secrets]:
{"WEBHOOK_USERNAME": "XXX","WEBHOOK_PASSWORD": "XXX"}

09/19/2021 13:34:09 - Checking for authorization ...

09/19/2021 13:34:09 - [Expected Authorization Header]:
Basic XXX

09/19/2021 13:34:09 - [Headers]:

Name                           Value
----                           -----
Content-Length                 {57}
User-Agent                     {curl/7.64.1}
Accept                         {*/*}

> SendToBroker                   {False}

Reviewer: Can we please change this from a header variable into a configuration option via ENV/secret configuration?

Host                           {localhost:8080}
Authorization                  {Basic XXX}
Content-Type                   {application/json}

09/19/2021 13:34:09 - [Body]:

> @{VmName=Production-01; UserName=Michael Gasch}

Reviewer: Can we use a fake user name please :)

> 09/19/2021 13:34:09 - Creating custom CloudEvent ...

Reviewer: nit: saying "custom" CloudEvent feels redundant, i.e. "CloudEvent" is just fine (see other occurrences).

> 09/19/2021 13:34:09 - Local developmen detected, will not forward custom CloudEvent to VMware Event Broker

Reviewer: ```suggestion
09/19/2021 13:34:09 - Running in local development: not sending CloudEvent to VMware Event Broker
```

```

# Step 3 - Deploy

> **Note:** The following steps assume a working Knative environment using the
`default` Rabbit `broker`. The Knative `service` and `trigger` will be installed in the
`vmware-functions` Kubernetes namespace, assuming that the `broker` is also available there.

Push your container image to an accessible registry such as Docker once you're done developing and testing your function logic.

```console
docker push <docker-username>/kn-ps-webhook:${TAG}
```

> Update the `webhook_secret.json` file with your vCenter Server credentials and configurations and then create the kubernetes secret which can then be accessed from within the function by using the environment variable named called `WEBHOOK_SECRET`.

Reviewer: The description feels like it's copy&paste and needs some adjustments for this example.

```yaml
annotations:
  autoscaling.knative.dev/maxScale: "1"
  autoscaling.knative.dev/minScale: "1"
```

# Step 4 - Verify webhook function

Retrieve the Knative service URL for the deployed function. In the example below, the name of the function is called `kn-ps-webhook` and ensure the the `READY` field shows `True`.

```console
# kubectl -n vmware-functions get ksvc kn-ps-webhook

NAME            URL                                                                 LATESTCREATED         LATESTREADY           READY   REASON
kn-ps-webhook   http://kn-ps-webhook.vmware-functions.veba.primp-industries.local   kn-ps-webhook-00001   kn-ps-webhook-00001   True
```

Change into the `test` directory
```console
cd test
```

Update the following variable names within the `send-incoming-webhook-to-veba.sh` script:

* WEBHOOK_USERNAME - Username to login to webhook function endpoint
* WEBHOOK_PASSWORD - Password to login to webhook function endpoint

> * WEBHOOK_FUNCTION_URL - Knative service URL for deployed webhook function

Reviewer: ```suggestion
* WEBHOOK_FUNCTION_URL - Knative service URL for deployed webhook function (retrieved in step above)
```

# Step 1 - Build

Create the container image locally to test your function logic.

```console
export TAG=<version>
docker build -t <docker-username>/kn-ps-webhook:${TAG} .
```

# Step 2 - Test

Verify the container image works by executing it locally.

Change into the `test` directory
```console
cd test
```

Update the following variable names within the `docker-test-env-variable` file and either the `send-cloudevent-test.ps1` (PowerShell Script) or `send-cloudevent-test.sh` (Bash Script)

* WEBHOOK_USERNAME - Username to login to webhook function endpoint
* WEBHOOK_PASSWORD - Password to login to webhook function endpoint
* WEBHOOK_CE_EVENT_TYPE - The type value used for creating the CloudEvent (e.g. com.vmware.generic.webhook.v0)
* WEBHOOK_SINK_URL - The sink URL (e.g. http://default-broker-ingress.vmware-functions.svc.cluster.local) to send the CloudEvent (leave blank for local development)

Start the container image by running the following command:

```console
docker run -e SERVICE_DEBUG=true -e PORT=8080 --env-file docker-test-env-variable -it --rm -p 8080:8080 <docker-username>/kn-ps-webhook:${TAG}
```

> In a separate terminal, run either `send-cloudevent-test.ps1` (PowerShell Script) or `send-cloudevent-test.sh` (Bash Script) to simulate a payload being sent to the local container image

Reviewer: Should we rename the test files to something like send-webhook... because you don't send a CloudEvent
